pkg/insights/insightsreport/insightsreport.go

```diff
@@ -0,0 +1,184 @@
+	insightsStatus = metrics.NewGaugeVec(&metrics.GaugeOpts{
+		Namespace: "health",
+		Subsystem: "statuses",
+		Name:      "insights",
+		Help:      "Information about the cluster health status as detected by Insights tooling.",
+	}, []string{"metric"})
+)
+
+// New initializes and returns a Gatherer
+func New(client *insightsclient.Client, configurator Configurator, reporter InsightsReporter) *Gatherer {
+	return &Gatherer{
+		Simple:                controllerstatus.Simple{Name: "insightsreport"},
+		configurator:          configurator,
+		client:                client,
+		archiveUploadReporter: reporter.ArchiveUploaded(),
+		retries:               0,
+	}
+}
+
+// PullSmartProxy performs a request to the Smart Proxy and unmarshal the response
+func (r *Gatherer) PullSmartProxy() (bool, error) {
+	klog.Info("Pulling report from smart-proxy")
+	config := r.configurator.Config()
+	initialWait := config.ReportPullingDelay
+	minimalRetryTime := config.ReportMinRetryTime
+	reportEndpoint := config.ReportEndpoint
+
+	if len(reportEndpoint) == 0 {
+		klog.V(4).Info("Not downloading report because Smart Proxy client is not properly configured: missing report endpoint")
+		return true, nil
+	}
+
+	delay := initialWait - time.Duration(r.retries*minimalRetryTime.Nanoseconds()) - minimalRetryTime
+	if delay > 0 {
+		time.Sleep(delay)
+	}
+	r.retries++
+
+	ctx, cancelFunc := context.WithTimeout(context.Background(), time.Minute)
+	defer cancelFunc()
+
+	klog.V(4).Info("Retrieving report")
+	reportBody, err := r.client.RecvReport(ctx, reportEndpoint)
+	if authorizer.IsAuthorizationError(err) {
+		r.Simple.UpdateStatus(controllerstatus.Summary{
+			Operation: controllerstatus.DownloadingReport,
+			Reason:    "NotAuthorized",
+			Message:   fmt.Sprintf("Auth rejected for downloading latest report: %v", err),
+		})
+		return false, err
+	} else if err != nil {
+		klog.Errorf("Error retrieving the report: %s", err)
+		return false, err
+	}
+
+	klog.V(4).Info("Report retrieved")
+	reportResponse := Response{}
+
+	if err = json.NewDecoder(*reportBody).Decode(&reportResponse); err != nil {
+		klog.Error("The report response cannot be parsed")
+		return false, err
+	}
+
+	klog.V(4).Info("Smart Proxy report correctly parsed")
+
+	if r.LastReport.Meta.LastCheckedAt == reportResponse.Report.Meta.LastCheckedAt {
+		klog.V(2).Info("Retrieved report is equal to previus one. Retrying...")
+		return false, fmt.Errorf("Report not updated")
+	}
+
+	updateInsightsMetrics(reportResponse.Report)
+	r.LastReport = reportResponse.Report
+	// close(r.insightsReport)
+	return true, nil
+}
+
+// Run goroutine code for gathering the reports from Smart Proxy
+func (r *Gatherer) Run(ctx context.Context) {
+	r.Simple.UpdateStatus(controllerstatus.Summary{Healthy: true})
+	klog.V(2).Info("Starting report retriever")
+
+	for {
+		// always wait for new uploaded archive or insights-operator ends
+		select {
+		case <-r.archiveUploadReporter:
```

Comment:
another reaction should be r.configurator.ConfigChanged(). Ideally it would keep some minimal refresh rate, which could be hard to implement using this pattern and insightsuploader has better tools for it.

Reply:
I get it. I'll keep working on that.

Reply:
New version includes a total refactor of the "polling loop". Instead of using one of the commodities in `wait` package, I implemented a new way that is aware of configuration changes.
The problem with configuration changes is that almost every config change will affect to the current timers, that are already set using `wait` functions.